rspace/src/test/scala/coop/rchain/rspace/ConcurrencyTests.scala

```diff
@@ -0,0 +1,110 @@
+package coop.rchain.rspace
+
+import coop.rchain.rspace.examples.AddressBookExample._
+import coop.rchain.rspace.examples.AddressBookExample.implicits._
+import coop.rchain.rspace.extended._
+
+trait ConcurrencyTests extends StorageTestsBase[Channel, Pattern, Entry, EntriesCaptor] {
+
+  def version: String
+
+  "CORE-589 produce and consume in a multithreaded fashion" should
+    "show high performance" in withTestStore { store =>
+    val arrResults      = new Array[Long](Runtime.getRuntime.availableProcessors)
+    val iterationsCount = 1000
+
+    def createThread(threadIndex: Int, iterations: Int): Thread =
+      new Thread(() => {
+
+        val channelName = "friends#" + threadIndex.toString
+        val start       = System.nanoTime()
+
+        for (iteration <- 1 to iterations) {
+          val r1 = consume(
+            store,
+            List(Channel(channelName), Channel(channelName)),
+            List(CityMatch(city = "Crystal Lake"), CityMatch(city = "Crystal Lake")),
+            new EntriesCaptor,
+            persist = false
+          )
+
+          r1 shouldBe None
+
+          val r2 = produce(store, Channel(channelName), bob, persist = false)
+
+          r2 shouldBe None
+
+          val r3 = produce(store, Channel(channelName), bob, persist = false)
+
+          r3 shouldBe defined
```

Comment:
I ran those tests on my machine and the assertions are failing:
```
sbt:rspace> testOnly *Concurrency*
[info] Compiling 9 Scala sources to /home/lukasz/git/pyrofex/rchain/rspace/target/scala-2.12/classes ...
[info] Done compiling.
[info] Compiling 1 Scala source to /home/lukasz/git/pyrofex/rchain/rspace/target/scala-2.12/classes ...
[info] Done compiling.
[info] Compiling 16 Scala sources to /home/lukasz/git/pyrofex/rchain/rspace/target/scala-2.12/test-classes ...
[warn] /home/lukasz/git/pyrofex/rchain/rspace/src/test/scala/coop/rchain/rspace/StorageTestsBase.scala:36:58: parameter value sp in class InMemoryStoreTestsBase is never used
[warn]                                                          sp: Serialize[P],
[warn]                                                          ^
[warn] /home/lukasz/git/pyrofex/rchain/rspace/src/test/scala/coop/rchain/rspace/StorageTestsBase.scala:37:58: parameter value sa in class InMemoryStoreTestsBase is never used
[warn]                                                          sa: Serialize[A],
[warn]                                                          ^
[warn] two warnings found
[info] Done compiling.
[info] Compiling 6 Scala sources to /home/lukasz/git/pyrofex/rchain/rspace/target/scala-2.12/test-classes ...
[info] Done compiling.
SLF4J: A number (2) of logging calls during the initialization phase have been intercepted and are
SLF4J: now being replayed. These are subject to the filtering rules of the underlying logging system.
SLF4J: See also http://www.slf4j.org/codes.html#replay
Exception in thread "Thread-21" org.scalatest.exceptions.TestFailedException: None was not defined
        at org.scalatest.MatchersHelper$.indicateFailure(MatchersHelper.scala:346)
        at org.scalatest.Matchers$AnyShouldWrapper.shouldBe(Matchers.scala:7073)
        at coop.rchain.rspace.ConcurrencyTests.$anonfun$$init$$4(ConcurrencyTests.scala:39)
        at coop.rchain.rspace.ConcurrencyTests.$anonfun$$init$$4$adapted(ConcurrencyTests.scala:22)
        at scala.collection.immutable.Range.foreach(Range.scala:156)
        at coop.rchain.rspace.ConcurrencyTests.$anonfun$$init$$3(ConcurrencyTests.scala:22)
        at java.lang.Thread.run(Thread.java:748)
Exception in thread "Thread-22" org.scalatest.exceptions.TestFailedException: None was not defined
        at org.scalatest.MatchersHelper$.indicateFailure(MatchersHelper.scala:346)
        at org.scalatest.Matchers$AnyShouldWrapper.shouldBe(Matchers.scala:7073)
        at coop.rchain.rspace.ConcurrencyTests.$anonfun$$init$$4(ConcurrencyTests.scala:39)
        at coop.rchain.rspace.ConcurrencyTests.$anonfun$$init$$4$adapted(ConcurrencyTests.scala:22)
        at scala.collection.immutable.Range.foreach(Range.scala:156)
        at coop.rchain.rspace.ConcurrencyTests.$anonfun$$init$$3(ConcurrencyTests.scala:22)
        at java.lang.Thread.run(Thread.java:748)
Exception in thread "Thread-27" org.scalatest.exceptions.TestFailedException: None was not defined
        at org.scalatest.MatchersHelper$.indicateFailure(MatchersHelper.scala:346)
        at org.scalatest.Matchers$AnyShouldWrapper.shouldBe(Matchers.scala:7073)
        at coop.rchain.rspace.ConcurrencyTests.$anonfun$$init$$4(ConcurrencyTests.scala:39)
        at coop.rchain.rspace.ConcurrencyTests.$anonfun$$init$$4$adapted(ConcurrencyTests.scala:22)
        at scala.collection.immutable.Range.foreach(Range.scala:156)
        at coop.rchain.rspace.ConcurrencyTests.$anonfun$$init$$3(ConcurrencyTests.scala:22)
        at java.lang.Thread.run(Thread.java:748)
Finished InMemory: 1000 iterations.
```

Reply:
The tests are nevertheless green, because assertion errors were thrown on threads other than the runner thread. I'm not sure if testing correctness and performance should be done in one place. I would tend to use JMH for benchmarks

Reply:
Hey, thanks for running tests on your machine. We definitely need to test correctness inside multi-threading test - otherwise we will not catch synchronization-related errors.
Seems on your machine the InMemory implementation doesn't work correctly.
What about running the concurrency test on LMDBStore? Is it running without errors or also fails?

Reply:
1) I'm only saying we shouldn't do it in one place, I'd split this into two tests - benchmarks and correctness.
2) `LMDBStore` seems to be working fine

Reply:
@henrytill - do we need to fix InMemoryStore to pass this test, or just remove it at all, what do you think?

@lukasz-golebiewski sounds reasonable, when I finish merge with latest dev, I will split pure performance test and test+correctness check and compare output. If there will be at least 0.2ms difference - will make them separate.